aspects/schema.go

```diff
@@ -0,0 +1,358 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+type parser interface {
+	Schema
+
+	// parseConstraints parses constraints for a type defined as a JSON object.
+	// Shouldn't be used with non-object/map type definitions.
+	parseConstraints(map[string]json.RawMessage) error
+}
+
+// ParseSchema parses a JSON aspect schema and returns a Schema that can be
+// used to validate aspects.
+func ParseSchema(raw []byte) (*StorageSchema, error) {
+	var schemaDef map[string]json.RawMessage
+	err := json.Unmarshal(raw, &schemaDef)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse top level schema: must be a map")
+	}
+
+	if rawType, ok := schemaDef["type"]; ok {
+		var typ string
+		if err := json.Unmarshal(rawType, &typ); err != nil {
+			return nil, fmt.Errorf(`cannot parse top level schema's "type" entry: %w`, err)
+		}
+
+		if typ != "map" {
+			return nil, fmt.Errorf(`cannot parse top level schema: expected map but got %s`, typ)
+		}
+	}
+
+	if _, ok := schemaDef["schema"]; !ok {
+		return nil, fmt.Errorf(`cannot parse top level schema: must have a "schema" constraint`)
+	}
+
+	// TODO: check "types" here and parse the user-defined types
+	schema := &StorageSchema{}
+	schema.topLevel, err = schema.parse(raw)
+	if err != nil {
+		return nil, err
```

Comment:
does this error need a prefix? (same for other places calling .parse) or should parse() add prefixes to more error?

Reply:
Yes, I'll add one. Once there are more constraints in place and this has filled out more, I'll do the work of keeping the path prefixes in the schemas and improve the errors so they're more consistent and informative

Reply:
I added them where the errors originate so we don't end up with things like: `cannot parse schema: cannot parse map: cannot parse map key constraint: cannot parse string: ...`